rfc/src/rfcs/0012-loop-contracts.md

```diff
@@ -0,0 +1,235 @@
+guard comes from the fact that a path exits the loop must fail the loop guard.
+After the loop is abstracted, the program will be equivalent to the following:
+```rs
+fn main() {
+    let mut x: u64 = kani::any_where(|i| *i >= 1);
+
+    x = kani::any(); // Arbitrary program state that
+    kani::assume( !(x > 1) && x >= 1); // satisfies !`guard` && `inv` 
+
+    assert!(x == 1);
+}
+```
+The assumption above is actually equivalent to `x == 1`, hence the assertion at the end
+of the program is proved.
+
+### Write Sets and Havocking
+
+For those memory locations that are not modified in the loop, loop invariants state
+that they stay unchanged throughout the loop are inductive. In other words, Kani should
+only havoc the memory locations that are modified in the loop. This is achieved by
+specifying the `modifies` clause for the loop. For example, the following program:
+```rs
+fn main() {
+    let mut x: u64 = kani::any_where(|i| *i >= 1);
+    let mut y: u64 = kani::any_where(|i| *i >= 1);
+
+    #[kani::loop_invariant(x >= 1)]
+    #[kani::loop_modifies(x)]
+    while x > 1{
+        x = x - 1;
+    };
+
+    assert!(x == 1);
+}
+```
+write to only `x` in the loop, hence the `modifies` clause contains only `x`.
+Then when use the loop contracts to abstract the loop, Kani will only havoc the memory
+location `x` and keep `y` unchanged. 
+
+
+Kani will also verify if all writing targets in the loop are included in the `modifies`
+clause.
+
+
+Note that the `modifies` clause is optional, and Kani will infer the write set if not
+provided.
+
+## Detailed Design
+
+
+Kani implements the functionality of loop contracts in three places.
+
+1. Procedural macros `loop_invariant` and `loop_modifies`.
+2. Code generation for builtin functions expanded from the above two macros.
+3. GOTO-level loop contracts using CBMC's contract language generated in
+   `kani-compiler` for `loop-modifies` clauses.
+
+### Procedural macros `loop_invariant` and `loop_modifies`.
+
+The `loop_invariant` macro perform code generation for the loop invariant clause.
+The generated code consists of two parts: 
+1. a closure definition to wrap the loop invariant, which is an boolean expression.
+2. a call to a builtin function `kani_loop_invariant` at end of the loop.
+
+As an example, in the above program, the following code will be generated for the
+loop invariants clauses:
+
+```rs
+fn main() {
+    let mut x: u64 = kani::any_where(|i| *i >= 1);
+
+    let _loop_invariant_closure = || x >= 1;
+    while x > 1{
+        x = x - 1;
+        kani_loop_invariant_begin();
+        _loop_invariant_closure()
+        kani_loop_invariant_end();
+    };
+
+    assert!(x == 1);
+}
+```
+
+Similarly, we generate a call to the builtin function `kani_loop_modifies` for modifies
+clauses.
+
+
+### Code Generation for Builtin Functions
+
+When generating GOTO program from MIR, Kani will first scan for the placeholder function
```

Comment:
Our codegen generate codes for each instruction at a time. Are you proposing to somehow change that? Is that really necessary?

Reply:
We can discuss implementation details a bit later, but I'm not convinced that this is the best approach since it adds more complexity to the codegen code.